src/low-level/smtp/mailsmtp.c

```diff
@@ -448,6 +448,33 @@ int mailsmtp_data_message(mailsmtp * session,
   }
 }
 
+/* lmtp data command
+ * on end of data transaction lmtp return a status for
+ * every recipient it delivered. the status is stored in
+ * a carray retcodes. the array must have the size of recips */
```

Comment:
its important for me to know which one failed to avoid duplicates, but I could make it optional for the caller. 
Problem for me is the read_response() in  line 472 in that case.
 If I dont know the count of responses, I dont know how often I have to call read_response() and if I call it without any response left, its waiting on a select() called from mailstream_read_line_append() until it timeouts.

I will change it to maillmtp_data_message(session, message, size,recipient_list, &retcodes);
with recipient_list needed and the retcodes pointer optional. if you have a nice way to avoid the read_response() behaviour I can make the recipient_list optional too.